### PR TITLE
Fix some exception messages for BindHandle-like cases

### LINKED ISSUE #66273: Test failure baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid6\\bindhandleinvalid6.cmd
Run: [runtime-coreclr outerloop 20220306.2](https://dev.azure.com/dnceng/public/_build/results?buildId=1647830&view=ms.vss-test-web.build-test-results-tab&runId=45502822&resultId=100064&paneView=debug)

Failed test:
```
R2R-CG2 windows arm64 Checked no_tiered_compilation @ Windows.10.Arm64v8.Open

- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid6\\bindhandleinvalid6.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid5\\bindhandleinvalid5.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid4\\bindhandleinvalid4.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid3\\bindhandleinvalid3.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid\\bindhandleinvalid.cmd

R2R-CG2 windows arm64 Checked @ Windows.10.Arm64v8.Open

- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid6\\bindhandleinvalid6.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid4\\bindhandleinvalid4.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid5\\bindhandleinvalid5.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid3\\bindhandleinvalid3.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid\\bindhandleinvalid.cmd

R2R-CG2 windows arm Checked @ Windows.10.Arm64v8.Open

- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid6\\bindhandleinvalid6.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid4\\bindhandleinvalid4.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid5\\bindhandleinvalid5.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid3\\bindhandleinvalid3.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid\\bindhandleinvalid.cmd

CoreCLR windows x86 Checked no_tiered_compilation @ Windows.10.Amd64.Open

- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid6\\bindhandleinvalid6.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid4\\bindhandleinvalid4.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid5\\bindhandleinvalid5.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid3\\bindhandleinvalid3.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid\\bindhandleinvalid.cmd


R2R-CG2 windows x64 Checked @ Windows.10.Amd64.Open

- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid6\\bindhandleinvalid6.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid4\\bindhandleinvalid4.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid5\\bindhandleinvalid5.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid3\\bindhandleinvalid3.cmd
- baseservices\\threading\\threadpool\\bindhandle\\bindhandleinvalid\\bindhandleinvalid.cmd
```
**One of error message:**
```
Could Not Find D:\h\w\B06A09A6\w\A88D0A0C\e\baseservices\threading\threadpool\bindhandle\bindhandleinvalid6\IL-CG2\composite-r2r.dll
Could Not Find D:\h\w\B06A09A6\w\A88D0A0C\e\baseservices\threading\threadpool\bindhandle\bindhandleinvalid6\bindhandleinvalid6.dll.rsp

Return code:      1
Raw output file:      D:\h\w\B06A09A6\w\A88D0A0C\uploads\Reports\baseservices.threading\threadpool\bindhandle\bindhandleinvalid6\bindhandleinvalid6.output.txt
Raw output:
BEGIN EXECUTION
bindhandleinvalid6.dll
1 file(s) copied.
Response file: D:\h\w\B06A09A6\w\A88D0A0C\e\baseservices\threading\threadpool\bindhandle\bindhandleinvalid6\bindhandleinvalid6.dll.rsp
D:\h\w\B06A09A6\w\A88D0A0C\e\baseservices\threading\threadpool\bindhandle\bindhandleinvalid6\IL-CG2\bindhandleinvalid6.dll
-o:D:\h\w\B06A09A6\w\A88D0A0C\e\baseservices\threading\threadpool\bindhandle\bindhandleinvalid6\bindhandleinvalid6.dll
--targetarch:arm64
--verify-type-and-field-layout
--method-layout:random
-r:D:\h\w\B06A09A6\p\System..dll
-r:D:\h\w\B06A09A6\p\Microsoft..dll
-r:D:\h\w\B06A09A6\p\mscorlib.dll
-r:D:\h\w\B06A09A6\p\netstandard.dll
-O
" "dotnet" "D:\h\w\B06A09A6\p\crossgen2\crossgen2.dll" @"D:\h\w\B06A09A6\w\A88D0A0C\e\baseservices\threading\threadpool\bindhandle\bindhandleinvalid6\bindhandleinvalid6.dll.rsp"   -r:D:\h\w\B06A09A6\w\A88D0A0C\e\baseservices\threading\threadpool\bindhandle\bindhandleinvalid6\IL-CG2*.dll"
Emitting R2R PE file: D:\h\w\B06A09A6\w\A88D0A0C\e\baseservices\threading\threadpool\bindhandle\bindhandleinvalid6\bindhandleinvalid6.dll
"D:\h\w\B06A09A6\p\corerun.exe" -p "System.Reflection.Metadata.MetadataUpdater.IsSupported=false"  bindhandleinvalid6.dll
BindHandle call succeeded
Got wrong error: System.ApplicationException: Error in the application.
at System.Threading.PortableThreadPool.RegisterForIOCompletionNotifications(IntPtr handle)
at System.Threading.ThreadPool.BindHandle(SafeHandle osHandle)
at BindHandle1.RunTest()
Expected: 100
Actual: 99
END EXECUTION - FAILED
FAILED
Test Harness Exitcode is : 1
To run the test:

set CORE_ROOT=D:\h\w\B06A09A6\p
D:\h\w\B06A09A6\w\A88D0A0C\e\baseservices\threading\threadpool\bindhandle\bindhandleinvalid6\bindhandleinvalid6.cmd
Expected: True
Actual:   False


Stack trace
   at baseservices_threading._threadpool_bindhandle_bindhandleinvalid6_bindhandleinvalid6_._threadpool_bindhandle_bindhandleinvalid6_bindhandleinvalid6_cmd()
```

### LINKED ISSUE #66274: Test failure baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh
Run: [runtime-coreclr outerloop 20220306.2](https://dev.azure.com/dnceng/public/_build/results?buildId=1647830&view=ms.vss-test-web.build-test-results-tab&runId=45502724&resultId=100101&paneView=debug)

Failed test:
```
R2R-CG2 Linux x64 Checked no_tiered_compilation @ Ubuntu.1804.Amd64.Open
- baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh

CoreCLR Linux x64 Checked no_tiered_compilation @ (Alpine.314.Amd64.Open)Ubuntu.1804.Amd64.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-helix-amd64-20210910135833-1848e19
- baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh

CoreCLR Linux x64 Checked @ (Alpine.314.Amd64.Open)Ubuntu.1804.Amd64.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-helix-amd64-20210910135833-1848e19
- baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh

R2R-CG2 Linux x64 Checked @ Ubuntu.1804.Amd64.Open
- baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh

R2R-CG2 Linux x64 Checked @ (Alpine.314.Amd64.Open)Ubuntu.1804.Amd64.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-helix-amd64-20210910135833-1848e19
- baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh

R2R-CG2 Linux arm64 Checked no_tiered_compilation @ (Ubuntu.1804.Arm64.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm64v8-20210531091519-97d8652

- baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh

R2R-CG2 Linux arm64 Checked no_tiered_compilation @ (Alpine.314.Arm64.Open)Ubuntu.1804.ArmArch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-helix-arm64v8-20210910135810-8a6f4f3
- baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh

R2R-CG2 Linux x64 Checked no_tiered_compilation @ (Alpine.314.Amd64.Open)Ubuntu.1804.Amd64.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-helix-amd64-20210910135833-1848e19
- baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh

CoreCLR Linux arm Checked no_tiered_compilation @ (Ubuntu.1804.Arm32.Open)Ubuntu.1804.Armarch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:ubuntu-18.04-helix-arm32v7-bfcd90a-20200121150440
- baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh

CoreCLR Linux arm Checked no_tiered_compilation @ (Alpine.314.Arm32.Open)Ubuntu.1804.ArmArch.Open@mcr.microsoft.com/dotnet-buildtools/prereqs:alpine-3.14-helix-arm32v7-20210910135806-8a6f4f3
- baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh

CoreCLR OSX arm64 Checked @ OSX.1200.ARM64.Open
- baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh

```
**One of error message:**
```
rm: cannot remove 'IL-CG2/composite-r2r.dll': No such file or directory
/datadisks/disk1/work/B67B09B5/w/B2D90968/e/baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh: line 246: -r:/datadisks/disk1/work/B67B09B5/w/B2D90968/e/baseservices/threading/threadpool/bindhandle/bindhandlenull/IL-CG2/*.dll: No such file or directory
rm: cannot remove '/datadisks/disk1/work/B67B09B5/w/B2D90968/e/baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.dll.rsp': No such file or directory

Return code:      1
Raw output file:      /datadisks/disk1/work/B67B09B5/w/B2D90968/uploads/Reports/baseservices.threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.output.txt
Raw output:
BEGIN EXECUTION
in takeLock
/datadisks/disk1/work/B67B09B5/w/B2D90968/e/baseservices/threading/threadpool/bindhandle/bindhandlenull/IL-CG2/bindhandlenull.dll
Response file: /datadisks/disk1/work/B67B09B5/w/B2D90968/e/baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.dll.rsp
/datadisks/disk1/work/B67B09B5/w/B2D90968/e/baseservices/threading/threadpool/bindhandle/bindhandlenull/IL-CG2/bindhandlenull.dll
-o:/datadisks/disk1/work/B67B09B5/w/B2D90968/e/baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.dll
-r:/datadisks/disk1/work/B67B09B5/p/System..dll
-r:/datadisks/disk1/work/B67B09B5/p/Microsoft..dll
-r:/datadisks/disk1/work/B67B09B5/p/mscorlib.dll
--verify-type-and-field-layout
--targetarch:x64
-O
Running CrossGen2:  dotnet /datadisks/disk1/work/B67B09B5/p/crossgen2/crossgen2.dll @/datadisks/disk1/work/B67B09B5/w/B2D90968/e/baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.dll.rsp

Emitting R2R PE file: /datadisks/disk1/work/B67B09B5/w/B2D90968/e/baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.dll
in ReleaseLock
/datadisks/disk1/work/B67B09B5/p/corerun -p System.Reflection.Metadata.MetadataUpdater.IsSupported=false bindhandlenull.dll ''
Unexpected: System.PlatformNotSupportedException: This API is specific to the way in which Windows handles asynchronous I/O, and is not supported on this platform.
at System.Threading.ThreadPool.BindHandle(SafeHandle osHandle)
at BindHandleNull.RunTest()
Expected: 100
Actual: 98
END EXECUTION - FAILED
Test Harness Exitcode is : 1
To run the test:

set CORE_ROOT=/datadisks/disk1/work/B67B09B5/p
/datadisks/disk1/work/B67B09B5/w/B2D90968/e/baseservices/threading/threadpool/bindhandle/bindhandlenull/bindhandlenull.sh
Expected: True
Actual:   False


Stack trace
   at baseservices_threading._threadpool_bindhandle_bindhandlenull_bindhandlenull_._threadpool_bindhandle_bindhandlenull_bindhandlenull_sh()
```

### PR DESCRIPTION
**Pull request by kouvel**
- Updated `ThrowHelper.ThrowApplicationException` to get and include an error message
- CoreCLR seems to have special code for producing exception messages from HRs and differs a bit from `Marshal.GetExceptionForHR`, forked to throw from the native side in CoreCLR to retain the previous messages
- Enabled the disabled tests, disabled a `BindHandle` test on Unixes where it now throws `PNSE` before `ArgumentNullException`, kept some tests disabled for Mono since the tests check for specific parts of exception messages that are specific to CoreCLR's code

Fixes https://github.com/dotnet/runtime/issues/66273
Fixes https://github.com/dotnet/runtime/issues/66274

### CODE REVIEW

**Comment by marek-safar** [comment]
Is it worth keeping coreclr special behaviour?

**Comment by kouvel** [comment]
It seems to produce slightly different messages sometimes and includes the HR and an `E_INVALIDARG`-like string indicating what it is. Not sure it's worth keeping since the HR is available on the exception.

File: src/coreclr/vm/comthreadpool.cpp [hunk 1]
```diff
@@ -625,6 +625,19 @@ extern "C" BOOL QCALLTYPE ThreadPool_PerformGateActivities(INT32 cpuUtilization)
     return needGateThread;
 }
 
+extern "C" void QCALLTYPE ThreadPool_ThrowApplicationException(INT32 hr)
+{
+    QCALL_CONTRACT;
+
+    BEGIN_QCALL;
+
+    // This has special code to produce an error message for an HR that differs a bit from Marshal.GetExceptionForHR(). This is
```
Comment (mangod9) [inline src/coreclr/vm/comthreadpool.cpp:634]: Is it important to maintain message level compat here, or could the tests be updated to check the new message? Assuming its throwing the same exception type.

Comment (kouvel) [reply src/coreclr/vm/comthreadpool.cpp:634]: As we discussed offline, I'll go ahead and do the simple thing to use Marshal and remove the special casing for CoreCLR.

**Review by mangod9 — APPROVED** [review]

**Review by stephentoub — APPROVED** [review]

**Pull request merged**